File: impl/ts/check.ts
// Phase-3 conformance harness for the TypeScript backend.
//
// Mirrors conformance/chanlun-v1/example_phase3_check.py: for every fixture in
// the FROZEN chanlun-v1 manifest, run the TS pipeline on `input`, canonicalize
// to JSON with the SAME rules Python uses (sort_keys, separators=(',', ':'),
// ensure_ascii), SHA-256 the bytes, compare against expected_sha256.
//
// All 48 fixtures must pass byte-for-byte. A single divergence => non-zero exit.
// No fuzzy match, no float tolerance - SHA-equality is the law (§15).

import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { canonical } from "./src/canonical.js";
import { normalizeN } from "./src/normalize.js";
import { extractFractals } from "./src/fractal.js";
import { strokes as strokesStage } from "./src/strokes.js";
import { zhongshu as zhongshuStage } from "./src/zhongshu.js";
import { classify as classifyStage } from "./src/trend_type.js";
import { DMIN, runFull } from "./src/pipeline.js";

interface ManifestEntry {
  id: string;
  stage: string;
  file: string;
  input_sha256: string;
  expected_sha256: string;
  fixture_sha256: string;
}

interface Manifest {
  version: string;
  corpus_sha256: string;
  count: number;
  entries: ManifestEntry[];
}

function backend(stage: string, input: unknown): unknown {
  switch (stage) {
    case "normalize":
      return normalizeN(input as Parameters<typeof normalizeN>[0]);
    case "fractal":
      return extractFractals(input as Parameters<typeof extractFractals>[0]);
    case "stroke": {
      const i = input as { fractals: Parameters<typeof strokesStage>[0]; dmin: number };
      const { strokes, dispositions } = strokesStage(i.fractals, i.dmin);
      return { strokes, dispositions };
    }
    case "zhongshu": {
      const i = input as {
        elements: Parameters<typeof zhongshuStage>[0];
        zone: Parameters<typeof zhongshuStage>[1];
      };
      return zhongshuStage(i.elements, i.zone);
    }
    case "trend_type":
      return classifyStage(input as Parameters<typeof classifyStage>[0]);
    case "pipeline": {
      const i = input as {
        bars: Parameters<typeof runFull>[0];
        dmin?: number;
        zone?: Parameters<typeof runFull>[2];
      };
      return runFull(i.bars, i.dmin ?? DMIN, i.zone ?? "first3");
    }
    default:
      throw new Error(`unknown stage: ${stage}`);
  }
}

function sha256Hex(bytes: Uint8Array): string {
  const h = createHash("sha256");
  h.update(bytes);
  return h.digest("hex");
}

async function main(): Promise<number> {
  // Resolve conformance/chanlun-v1/ relative to the compiled check.js location,
  // walking up from impl/ts/dist/ to the repo root.
  const here = dirname(fileURLToPath(import.meta.url));
  // dist/check.js -> dist -> ts -> impl -> repo root
  const repoRoot = resolve(here, "..", "..", "..");
  const corpusDir = resolve(repoRoot, "conformance", "chanlun-v1");

  const manifestText = await readFile(resolve(corpusDir, "manifest.json"), "utf-8");
  const manifest = JSON.parse(manifestText) as Manifest;

  console.log("chanlun-v1 phase-3 check, backend = impl/ts");
  console.log(`  corpus_sha256 = ${manifest.corpus_sha256}`);

  let fails = 0;
  for (const entry of manifest.entries) {
    const fixtureText = await readFile(resolve(corpusDir, entry.file), "utf-8");
    const fixture = JSON.parse(fixtureText) as {
      stage: string;
      input: unknown;
      expected_sha256: string;
    };
    let actual: unknown;
    try {
      actual = backend(fixture.stage, fixture.input);
    } catch (e) {
      const msg = e instanceof Error ? `${e.name}: ${e.message}` : String(e);
      console.log(`  [FAIL] ${entry.id.padEnd(60)} backend raised ${msg}`);
      fails += 1;
      continue;
    }
    const bytes = new TextEncoder().encode(canonical(actual));
    const actSha = sha256Hex(bytes);
    if (actSha !== entry.expected_sha256) {
      console.log(
        `  [FAIL] ${entry.id.padEnd(60)} actual sha=${actSha.slice(0, 12)}... expected sha=${entry.expected_sha256.slice(0, 12)}...`
      );
      fails += 1;
    }
  }

  const total = manifest.count;
  console.log();
  console.log(`Result: ${total - fails} PASS, ${fails} FAIL`);
  return fails ? 1 : 0;
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(2);
  });

File: impl/java/src/main/java/com/minustwo/chanlun/check/Check.java
// Check.java — Java conformance harness for the chanlun-v1 frozen corpus.
//
// Mirrors conformance/chanlun-v1/example_phase3_check.py + impl/go/cmd/check/main.go:
//   1. Load conformance/chanlun-v1/manifest.json.
//   2. For each entry, load the fixture JSON, invoke the Java backend on `input`,
//      canonicalize the output via our canonical-JSON emitter
//      (sort_keys=True, separators=(",",":"), ensure_ascii=True — matching
//      Python json.dumps byte-for-byte), and compare SHA-256 against the entry.
//   3. Exit non-zero on ANY divergence.
//
// SHA-equality is the law. A byte off = FAIL. No fuzzy comparison, no float
// tolerance (the chanlun-v1 corpus is integer-core).
package com.minustwo.chanlun.check;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import com.minustwo.chanlun.Canonical;
import com.minustwo.chanlun.Pipeline;
import com.minustwo.chanlun.Types.OrderedObject;
import com.minustwo.chanlun.Types.Value;

public final class Check {

    private Check() {}

    public static void main(String[] args) {
        try {
            int code = run();
            System.exit(code);
        } catch (Exception e) {
            System.err.println("FAIL: " + e.getClass().getSimpleName() + ": " + e.getMessage());
            e.printStackTrace(System.err);
            System.exit(1);
        }
    }

    private static int run() throws IOException {
        Path manifestPath = locateManifest();
        byte[] manifestBytes = Files.readAllBytes(manifestPath);
        Value manV = Canonical.decode(manifestBytes);
        OrderedObject manObj = manV.mustObj();
        List<Value> entries = manObj.get("entries").mustArr();
        String corpusSha = manObj.get("corpus_sha256").mustStr();
        long count = manObj.get("count").mustInt();

        System.out.println("chanlun-v1 Java conformance check: " + count + " fixtures");
        System.out.println("  corpus_sha256 = " + corpusSha);

        // Corpus-level sanity: SHA-256(canonical([fixture_sha256 for each entry])).
        List<Value> fxs = new ArrayList<>(entries.size());
        for (Value e : entries) {
            fxs.add(Value.ofString(e.mustObj().get("fixture_sha256").mustStr()));
        }
        String recomputedCorpus = Canonical.sha256Hex(Canonical.canonical(Value.ofArray(fxs)));
        if (!recomputedCorpus.equals(corpusSha)) {
            System.err.println("FAIL: corpus_sha256 mismatch: manifest=" + corpusSha
                + " recomputed=" + recomputedCorpus);
            return 1;
        }

        Path manifestDir = manifestPath.getParent();
        int pass = 0, fail = 0;
        for (Value e : entries) {
            OrderedObject eo = e.mustObj();
            String id = eo.get("id").mustStr();
            String stage = eo.get("stage").mustStr();
            String file = eo.get("file").mustStr();
            String inShaExp = eo.get("input_sha256").mustStr();
            String expShaExp = eo.get("expected_sha256").mustStr();
            String fixShaExp = eo.get("fixture_sha256").mustStr();

            Path fxPath = manifestDir.resolve(file);
            byte[] fxBytes;
            try {
                fxBytes = Files.readAllBytes(fxPath);
            } catch (IOException ex) {
                fail++;
                System.out.printf("  [FAIL] %-60s  read fixture: %s%n", id, ex.getMessage());
                continue;
            }

            Value fxV;
            try {
                fxV = Canonical.decode(fxBytes);
            } catch (RuntimeException ex) {
                fail++;
                System.out.printf("  [FAIL] %-60s  decode fixture: %s%n", id, ex.getMessage());
                continue;
            }
            OrderedObject fo = fxV.mustObj();
            Value inputV = fo.get("input");
            Value expectedV = fo.get("expected");

            byte[] inCanon = Canonical.canonical(inputV);
            String inSha = Canonical.sha256Hex(inCanon);
            if (!inSha.equals(inShaExp)) {
                fail++;
                System.out.printf("  [FAIL] %-60s  input_sha256: manifest=%s got=%s%n",
                    id, inShaExp, inSha);
                continue;
            }

            Value actualV;
            try {
                actualV = Pipeline.runStage(stage, inputV);
            } catch (RuntimeException ex) {
                fail++;
                System.out.printf("  [FAIL] %-60s  backend: %s%n", id, ex.getMessage());
                continue;
            }
            byte[] expCanon = Canonical.canonical(expectedV);
            byte[] actCanon = Canonical.canonical(actualV);
            String expSha = Canonical.sha256Hex(expCanon);
            String actSha = Canonical.sha256Hex(actCanon);
            if (!expSha.equals(expShaExp)) {
                fail++;
                System.out.printf("  [FAIL] %-60s  expected_sha256: manifest=%s recomputed=%s%n",
                    id, expShaExp, expSha);
                continue;
            }
            if (!actSha.equals(expSha)) {
                fail++;
                System.out.printf("  [FAIL] %-60s  actual sha=%s expected sha=%s%n",
                    id, actSha, expSha);
                System.out.println("           expected canon=" + new String(expCanon));
                System.out.println("           actual   canon=" + new String(actCanon));
                continue;
            }
            // fixture_sha = sha256(in_canon + "|" + exp_canon)
            byte[] joined = new byte[inCanon.length + 1 + expCanon.length];
            System.arraycopy(inCanon, 0, joined, 0, inCanon.length);
            joined[inCanon.length] = (byte) '|';
            System.arraycopy(expCanon, 0, joined, inCanon.length + 1, expCanon.length);
            String fxSha = Canonical.sha256Hex(joined);
            if (!fxSha.equals(fixShaExp)) {
                fail++;
                System.out.printf("  [FAIL] %-60s  fixture_sha256: manifest=%s recomputed=%s%n",
                    id, fixShaExp, fxSha);
                continue;
            }
            pass++;
        }

        System.out.println();
        System.out.printf("Result: %d PASS, %d FAIL%n", pass, fail);
        if (fail > 0) {
            System.err.println("conformance FAILED: " + fail
                + " divergence(s) — any byte-level mismatch breaks the FROZEN spec");
            return 1;
        }
        System.out.println("Conformance OK. All fixtures match the chanlun-v1 frozen spec byte-for-byte.");
        return 0;
    }

    /** Walk up from CWD looking for conformance/chanlun-v1/manifest.json so the
     *  harness works whether invoked from impl/java or from the repo root. */
    private static Path locateManifest() {
        Path dir = Paths.get("").toAbsolutePath();
        while (dir != null) {
            Path cand = dir.resolve("conformance").resolve("chanlun-v1").resolve("manifest.json");
            if (Files.exists(cand)) return cand;
            dir = dir.getParent();
        }
        throw new IllegalStateException(
            "could not locate conformance/chanlun-v1/manifest.json from " + Paths.get("").toAbsolutePath());
    }
}
